### PR TITLE
drivers/location: Add location API

### PR DESCRIPTION
**Pull request by bjarki-andreasen**
**Purpose**
The Location API is designed to provide the essential location
data which can be produced from sources like SUPL. GNSS, WIFI-
location, cellular location, proprietary solutions etc.

**NOTE**
This is a continuation from the following closed PR https://github.com/zephyrproject-rtos/zephyr/pull/46660,
from which the location API has been altered to be nearly unrecognizable. This PR should allow for a more
efficient discussion regarding the newest design.

**Future**
It is primed for use with the upcoming multi interface device
model, and for implementation into the upcoming GNSS API.

### CODE REVIEW

File: include/zephyr/drivers/location.h [hunk 1]
```diff
@@ -0,0 +1,315 @@
...
+#define LOCATION_EVENT_POSITION_UPDATED BIT(0)
+#define LOCATION_EVENT_BEARING_UPDATED BIT(1)
+#define LOCATION_EVENT_SPEED_UPDATED BIT(2)
+#define LOCATION_EVENT_ALTITUDE_UPDATED BIT(3)
+
+/**
+ * @typedef location_event_handler_t
+ * @brief Callback API invoked when a location event is raised
+ *
+ * See struct location_provider_event for description of the event
```
Comment (ycsin) [inline include/zephyr/drivers/location.h:150]: Not sure if I missed anything, is `location_provider_event` defined anywhere?

File: subsys/location/Kconfig [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+	help
+	  Include location subsystem
+
+if LOCATION
+
+config LOCATION_PROVIDERS_MAX
+	int "Max supported providers"
+	default 4
+	help
+	  Sets the maximum supported providers
```
Comment (jukkar) [inline subsys/location/Kconfig]: I wonder if we could avoid specifying the max providers by utilizing the `STRUCT_SECTION_ITERABLE` macro that could add the providers into a special linker section and then we could just iterate through that section at runtime. This pattern is used in many places in zephyr and it avoids allocating any extra data that might not be used.

File: include/zephyr/drivers/location.h (hunk 1 above)
Comment (bjarki-trackunit) [reply include/zephyr/drivers/location.h]: That was a remnant from the old design, it has been removed, on top of this, a typedef for the events mask has been added location_events_t along with documentation associating LOCATION_EVENT_ ... with the event handler events argument :)

File: subsys/location/Kconfig (hunk 2 above)
Comment (bjarki-trackunit) [reply subsys/location/Kconfig]: The latest commit only changes 6 files, and does not include the subsystem at all :) That is from the previous design.

Comment (bjarki-trackunit) [reply subsys/location/Kconfig]: I have squashed the two commits so it it clearer :)

**Comment by mrclark25** [comment]
Is this still planned? Looks like the accompanying GNSS API PR was closed: https://github.com/zephyrproject-rtos/zephyr/pull/48553

**Comment by bjarki-trackunit** [comment]
> Is this still planned? Looks like the accompanying GNSS API PR was closed: #48553

Yes, it is part of a two GNSS drivers for the Quectel L76LL and Quectel LCX6G modems, which should be ready by next week. I redesigned the GNSS API substantially, which will be part of the GNSS drivers PR, so the old GNSS API is no longer relevant.,

**Comment by mrclark25** [comment]
> Yes, it is part of a two GNSS drivers for the Quectel L76LL and Quectel LCX6G modems, which should be ready by next week. I redesigned the GNSS API substantially, which will be part of the GNSS drivers PR, so the old GNSS API is no longer relevant.,

Thank you for the update!

**Comment by beriberikix** [comment]
I realize activity went stale on this issue but LBS would be a great enhancement to Zephyr!

**Comment by karstenes** [comment]
Is this still planned? I would be interested in assisting with the implementation of drivers for some GNSS modems. Is it recommended to just use a generic sensor API instead?

**Comment by fabiobaltieri** [comment]
@bjarki-trackunit are you still interested in working on this? There's certainly interest for this, thinking maybe proposing an event based implementation could gain traction, maybe just a callback with an event to get full location updates (position, speed, bearing etc...) and the subsystem level API may just be a set_attribute/get_attribute pair to configured generic and vendor specific parameters.

**Comment by bjarki-trackunit** [comment]
I ended up creating two APIs, one for location, and one for GNSS, using the functions in this API. From that experience, I found that all these parameters should be combined into one struct, as you also suggest. The accuracy of lat/long is quite useful, accuracy and uptime was never used. accuracy for speed/bearing/altitude where never used.

My quick proposal for the perfect location data type to be used in APIs that provide location (like GNSS):

Creating this data type
```
struct location {
    /** Latitudal position in nanodegrees (0 to +-180000000000) */
    int64_t latitude;
    /** Longitudal position in nanodegrees (0 to +-180000000000) */
    int64_t longitude;
    /** Accuracy radius in millimeters */
   uint32_t accuracy;
    /** Bearing angle relative to True North in millidegrees (0 to 360000) */
    uint32_t bearing;
    /** Speed over ground in millimeters per second */
    uint32_t speed;
    /** Altitude in millimeters */
    int32_t altitude;
};
```
I can create a PR with the GNSS API I ended up using, that uses this struct for location if that is of interest?

**Comment by fabiobaltieri** [comment]
> I ended up creating two APIs, one for location, and one for GNSS, using the functions in this API. From that experience, I found that all these parameters should be combined into one struct, as you also suggest. The accuracy of lat/long is quite useful, accuracy and uptime for speed/bearing/altitude where never used.
> 
> My quick proposal for the perfect location data type to be used in APIs that provide location (like GNSS):
> 
> ```
> struct location {
>     /** Latitudal position in nanodegrees (0 to +-180000000000) */
>     int64_t latitude;
>     /** Longitudal position in nanodegrees (0 to +-180000000000) */
>     int64_t longitude;
>     /** Accuracy radius in millimeters */
>     uint32_t accuracy;
>     /** Bearing angle relative to True North in millidegrees (0 to 360000) */
>     uint32_t bearing;
>     /** Speed over ground in millimeters per second */
>     uint32_t speed;
>     /** Altitude in millimeters */
>     int32_t altitude;
> };
> ```
> 
> I can create a PR with the GNSS API I ended up designing and implementing, that uses this struct for location if that is of interest?
> 
> Maybe the struct is too big though

I think it may be fine (and convenient) to have a single big structure if it does not get copied around at subsystem level.  Maybe you could have a single structure allocated at the driver level, and then it gets passed by all the listeners and unlocked, so the main data passing would be a publisher-subscriber thing (either an iterable section like the bluetooth stuff, input etc... or zbus), then the `__subsystem` stuff would only deal with settings, maybe you could exclude specific messages if the application does not need them and stuff like that. And then if extensibility is a thing you could imagine an extra `void *` in the main location update structure that can be used for device specific data.

**Comment by bjarki-trackunit** [comment]
# Update

A continuation of this PR has been opened! #61073

**Pull request closed without merging**